src/X11R6/contrib/Motif/winterp/doc/wl.man
.\" $Header: /disk3/npm/src/winterp/doc/RCS/wl.man,v 2.4 1994/06/08 18:43:51 npm Exp $
.\" Copyright (C) 1994, Enterprise Integration Technologies Corp. and Niels Mayer.
.\" WINTERP 1.15-1.99, Copyright (c) 1993, Niels P. Mayer.
.\" WINTERP 1.0-1.14, Copyright (c) 1989-1992 Hewlett-Packard Co. and Niels Mayer.
.\" 
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation, and that the name of Enterprise Integration Technologies,
.\" Hewlett-Packard Company, or Niels Mayer not be used in advertising or
.\" publicity pertaining to distribution of the software without specific,
.\" written prior permission. Enterprise Integration Technologies, Hewlett-Packard
.\" Company, and Niels Mayer makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is" without express or
.\" implied warranty.
.\" 
.\" ENTERPRISE INTEGRATION TECHNOLOGIES, HEWLETT-PACKARD COMPANY AND NIELS MAYER
.\" DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL ENTERPRISE
.\" INTEGRATION TECHNOLOGIES, HEWLETT-PACKARD COMPANY OR NIELS MAYER BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
.\" RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
.\" CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"  
.TH WL 1 "June 6, 1994" "X11r6 contrib"
.SH NAME
WL - Winterp-Lisp client for communicating with WINTERP UN*X Domain Eval-Server.
.SH SYNOPSIS
.B wl
[\fB\-f <socket-filepath>\fP] 
[\fB\<s-expr>\fP]
.SH DESCRIPTION

\-\-\->> FOR DETAILS ON WINTERP SEE /usr/local/winterp/doc/winterp.doc <<\-\-\-
.PP
The WINTERP client program 
.I wl
(stands for Winterp-Lisp) sends the
XLISP expression given on its command line to the WINTERP server.
.I wl
hangs until WINTERP has evaluated the s\-expression given on
.I wl's
command-line. After the s\-expression is successfully
evaluated by WINTERP, 
.I wl
prints to stdout the results of the evaluation,
and sets exit status 0. If an evaluation error occurs, an error messsage is
output to stderr, and exit status gets set to 127.
.PP
By default, WINTERP compiles with the Unix-Domain eval server enabled. If
for some reason your machine or OS does not support Unix Domain Sockets,
you may turn off compilation of these features. Do this by (1) removing
"\-DWINTERP_WANT_UNIX_SERVER" in the associated
<winterp\-top\-dir>/src\-server/Makefile.* makefile; or (2) make equivalent
change to the Imakefile or its generated Makefile in that directory; or (3)
edit <winterp\-top\-dir>/src\-server/config.h...
.PP
The Unix Domain Socket can also be enabled/disabled via X resource
<<Winterp.enableUnixServer:>>, or via command\-line arguments
<<\-no_unix_server>> or <<\-enable_unix_server>>. 
.PP
.I wl
will "hang" waiting for WINTERP to finish the evaluation.
When the evaluation completes, 
.I wl
prints the results of the evaluation to
stdout, and returns exit status 0.  If an error occurs in evaluation, an
exit status of 128 is returned and the following is printed to stderr: "wl:
an evaluation error occured in WINTERP server."  Finally, if some other
error occurs, e.g. network error, unable to bind socket, etc, then exit
status of 1 is returned and an eror message will appear on stderr.
.PP
If evaluation via 
.I wl
causes an error and *BREAKENABLE* is set
(including by selecting "Error Break-Loop" or "Error Trace-Back"
in the "WINTERP Control Panel), then 
.I wl
will hang until the
break-loop exits, typically because the user has invoked functions
'clean\-up', 'top\-level', or 'continue', or typed Control\-C or
Control\-D into WINTERP's controlling terminal.
.PP
Example: invoking the following in a UN*X command shell:
.PP
		wl '(load "rc\-shell")'
.PP
while there is a WINTERP running with a unix domain socket enable, will
cause WINTERP to load the file "rc\-shell" and pop up the widgetry
associated with loading that file.
.SH OPTIONS

.BI [\-f <socket\-filepath>] 

The optional <socket\-filepath> argument sets the pathname for the Unix
Domain Socket eval\-server.  By default, <socket\-filepath> is
"/tmp/.winterp_serv", you may want to specify a different <socket\-filepath>
if multiple 
.I winterp
 applications are running. Alternately, you may want
to hide the Unix Domain Socket from other users if you are on a multi-user
machine. Note that you may specify <socket\-filepath> via environment
variable WINTERP_UNIX_SOCKET_FILEPATH. The environment variable sets
<socket\-filepath> for both 
.I wl
and 
.I winterp
.

.BI [<s\-expr>] 

The <s\-expr> argument is a single Lisp s\-expression to be sent to WINTERP's
Lisp listener. If no <s\-expr> argument is given, 
.I wl
sends an EOF
(==Control\-D) to XLISP, which causes WINTERP to terminate. Note that the
<s\-expr> must be quoted with ['] so that the shell does not attempt to
interpret the parentheses as shell commands.

.SH AVAILABILITY:
You may obtain the latest released version of the WINTERP source,
documentation, and examples via anonymous ftp from internet host ftp.x.org
in directory /contrib/devel_tools, file winterp\-2.xx.tar.gz, where 'xx'
represents the revision number. Directory /contrib/devel_tools/winterp
contains slides, papers, further documentation, contributed code, etc.
.PP
An automated reply indicating the locations and status of the latest
WINTERP source may be obtained by sending an e\-mail message to
winterp\-source@netcom.com.

.SH CONTACTING OTHER USERS
winterp@netcom.com is the mailing list for WINTERP-related announcements
and discussions. To get added/removed from the list, send mail to
winterp\-request@netcom.com. Please do not send junk mail or
subscribe/unsubscribe messages to the mailing list winterp@netcom.com.
.PP
For discussions about XLISP, see the USENET newsgroup comp.lang.lisp.x.
.SH SEE ALSO
winterp(1), wl\-tcpip(1).
.SH FILES
/tmp/.winterp_serv
$WINTERP_UNIX_SOCKET_FILEPATH
.SH AUTHOR
Niels Mayer (mayer@netcom.com)


src/NeWS
../NeWS

src/X11R6/contrib/Motif/winterp/src-client/wl.c
/* -*-C-*-
********************************************************************************
*
* File:         wl.c
* RCS:          $Header: /users/npm/src/winterp/src-client/RCS/wl.c,v 2.5 1994/06/06 15:14:40 npm Exp $
* Description:  Unix Domain CLIENT to send lisp expressions to WINTERP lisp-server.
*		This code derived from winterp/src-client/wl-tcpip.c
*		as well as code contributed by Victor Kan.
* Author:       Niels Mayer, Victor Kan <kan@DG-RTP.DG.COM>
* Created:      Sat Jun 10 02:15:35 1989
* Modified:     Sun Jun  5 16:12:42 1994 (Niels Mayer) npm@indeed
* Language:     C
* Package:      N/A
* Status:       X11r6 contrib release
*
* Copyright (C) 1994, Enterprise Integration Technologies Corp. and Niels Mayer.
* WINTERP 1.15-1.99, Copyright (c) 1993, Niels P. Mayer.
* WINTERP 1.0-1.14, Copyright (c) 1989-1992 Hewlett-Packard Co. and Niels Mayer.
* 
* Permission to use, copy, modify, distribute, and sell this software and its
* documentation for any purpose is hereby granted without fee, provided that
* the above copyright notice appear in all copies and that both that
* copyright notice and this permission notice appear in supporting
* documentation, and that the name of Enterprise Integration Technologies,
* Hewlett-Packard Company, or Niels Mayer not be used in advertising or
* publicity pertaining to distribution of the software without specific,
* written prior permission. Enterprise Integration Technologies, Hewlett-Packard
* Company, and Niels Mayer makes no representations about the suitability of
* this software for any purpose.  It is provided "as is" without express or
* implied warranty.
* 
* ENTERPRISE INTEGRATION TECHNOLOGIES, HEWLETT-PACKARD COMPANY AND NIELS MAYER
* DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL ENTERPRISE
* INTEGRATION TECHNOLOGIES, HEWLETT-PACKARD COMPANY OR NIELS MAYER BE LIABLE
* FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
* RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
* CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
* CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*
********************************************************************************
*/
static char rcs_identity[] = "@(#)$Header: /users/npm/src/winterp/src-client/RCS/wl.c,v 2.5 1994/06/06 15:14:40 npm Exp $";

/*
*------------------------------------------------------------------------------
* See ./winterp/COPYRIGHT for information on contacting the authors.
* Please e-mail comments, modifications, questions, improvements and
* bugfixes to the WINTERP mailing list winterp@netcom.com. Please send 
* mailing list subscribe/unsubscribe notices to winterp-request@netcom.com .
* Post XLISP-specific questions/information to the USENET newsgroup
* comp.lang.lisp.x.
*------------------------------------------------------------------------------
*/

#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include "../src-server/config.h"

char *progname = NULL;


/*******************************************************************************
 * Send <message> to the server via socket <s>
 *******************************************************************************/
void Server_Send(s, message)
     int s;
     char *message;
{
  if (send(s, message, strlen(message), 0) < 0) {
    perror(progname);
    fprintf(stderr, "%s: unable to send() on Unix Domain Socket.\n", progname);
    exit(1);
  }
}


/*******************************************************************************
 * Establish a connection with server process, returning socket descriptor
 * if successful.
 *
 * This is an AF_UNIX version of wl.c:Server_Connect()
 *******************************************************************************/
int Server_Connect(socket_path)
     char *socket_path;
{
  int s;
  struct sockaddr_un peeraddr_un;

  peeraddr_un.sun_family = AF_UNIX;

  /*
   * Set the file system entry used for the AF_UNIX Domain Socket.
   */
#ifndef SOCKADDR_UN_MAXLEN
#define SOCKADDR_UN_MAXLEN 108	/* can't find SOCKADDR_UN_MAXLEN on hpux 7.0, however "char sun_path[108];" */ 
#endif
  if (strlen(socket_path) > (SOCKADDR_UN_MAXLEN - 1)) {
    fprintf(stderr, "%s: Unix Domain Socket filepath %s must be shorter than %d bytes.\n",
	    progname,
	    socket_path,
	    SOCKADDR_UN_MAXLEN - 1);
    exit(1);
  }
  else
    strcpy(peeraddr_un.sun_path, socket_path);

  /*
   * Create the socket.
   */
  if ((s = socket(peeraddr_un.sun_family, SOCK_STREAM, 0)) == -1) {
    perror(progname);
    fprintf(stderr, "%s: socket() failed to create Unix Domain Socket %s .\n",
	    progname, socket_path);
    exit(1);
  }
  
  /*
   * Try to connect to the local server through the socket
   * which was just built into peeraddr.
   */
  if (connect(s, (struct sockaddr *) &peeraddr_un, sizeof(peeraddr_un.sun_family) + strlen(peeraddr_un.sun_path)) == -1) {
    perror(progname);
    fprintf(stderr, "%s: unable to connect() to remote via Unix Domain Socket %s .\n",
	    progname, socket_path);
    exit(1);
  }
  
  return (s);
}


/******************************************************************************
 * Server_Disconnect -- shutdown the connection for further sends. This causes
 * the server to receive an EOF condition after it has received all data from
 * Server_Send().
******************************************************************************/
void Server_Disconnect(s)
     int s;
{
  static char rdbuf[BUFSIZ];
  int len;
  
  Server_Send(s, WINTERP_CLIENT_SPECIAL_EOF_STRING); /* !!! hack !!! -- see ../src-server/winterp.c:Read_Eval_Print() and ../src-server/config.h */
  
  if (shutdown(s, 1) == -1) {
    perror(progname);
    fprintf(stderr, "%s: unable to shutdown() Unix Domain Socket.\n", progname);
    exit(1);
  }

  if ((len = recv(s, rdbuf, BUFSIZ, 0)) > 0) {
    rdbuf[len] = '\0';
    fprintf(stdout, "%s", rdbuf);
  }
  else if (len == 0) {
    fprintf(stderr, "%s: an evaluation error occured in WINTERP server.\n", progname);
    exit(128);			/* this sets exit status to "128" */
  }
  else if (len < 0) {
    perror(progname);
    fprintf(stderr, "%s: unable to recv() server reply from Unix Domain Socket.\n", progname);
    exit(1);			/* this sets exit status to "1" */
  }

  while ((len = recv(s, rdbuf, BUFSIZ, 0)) > 0) {
    rdbuf[len] = '\0';
    fprintf(stdout, "%s", rdbuf);
  }

  fprintf(stdout, "\n");

  if (len < 0) {
    perror(progname);
    fprintf(stderr, "%s: unable to recv() server reply from Unix Domain Socket.\n", progname);
    exit(1);			/* this sets exit status to "1" */
  }
  else {
    exit(0);			/* this sets exit status to "0" */
  }
}


/******************************************************************************
 * AF_UNIX version of wl-tcpip.c:main()
 ******************************************************************************/
main(argc, argv)
     int argc;
     char *argv[];
{
  extern int getopt();		/* proc to get option letter from argv */
  extern char* optarg;		/* arg string being pointed to by getopt() */
  extern int optind;		/* index into argv set by getopt() */
  int opt;			/* option argument */
  char* socket_path = NULL;	/* for pointing to environment vars */
  int s;			/* connected socket desriptor */

  progname = argv[0];

  while ((opt = getopt(argc, argv, "f:")) != EOF)
    switch (opt) {
    case 'f':
      socket_path = (char*) malloc((unsigned) strlen(optarg) + 1);
      strcpy(socket_path, optarg);
      break;
    case '?':
      fprintf(stderr, "usage: %s [-f <unix-socket-filepath>] [sexpr]\n", progname);
      fprintf(stderr, "\t<unix-socket-filepath> defaults to %s\n", DEFAULT_UNIX_SOCKET_FILEPATH); /* DEFAULT_UNIX_SOCKET_FILEPATH in "../src-server/config.h" */
      exit(1);
      break;
    }

  if (socket_path == NULL)
    if ((socket_path = (char *) getenv(DEFAULT_UNIX_SOCKET_FILEPATH_ENVVAR)) == NULL) /* DEFAULT_UNIX_SOCKET_FILEPATH_ENVVAR in "../src-server/config.h" */
      socket_path = DEFAULT_UNIX_SOCKET_FILEPATH;

  s = Server_Connect(socket_path);
    
  /*
   * note that due to a bug/feature of winterp, we must send it only
   * one sexp. Here, in the case that multiple args get passed, we also
   * want to ensure that they're whitespace separated for tokenization
   */
  for (; optind < argc; optind++) {
    Server_Send(s, argv[optind]);
    Server_Send(s, " ");
  }

  Server_Disconnect(s);		/* this routine calls exit() */
}
